tests/test_series.py

```diff
@@ -3276,18 +3276,36 @@ def test_map_na() -> None:
 
     mapping = {1: "a", 2: "b", 3: "c"}
     check(assert_type(s.map(mapping, na_action=None), "pd.Series[str]"), pd.Series, str)
+    check(assert_type(s.map(mapping), "pd.Series[str]"), pd.Series, str)
 
     def callable(x: int | NAType) -> str | NAType:
         if isinstance(x, int):
             return str(x)
         return x
 
+    def bad_callable(x: int) -> int:
+        return x << 1
+
+    with pytest.raises(TypeError):
+        s.map(
+            bad_callable, na_action=None  # type: ignore[arg-type] # pyright: ignore[reportCallIssue, reportArgumentType]
+        )
+    with pytest.raises(TypeError):
+        s.map(bad_callable)  # type: ignore[arg-type] # pyright: ignore[reportArgumentType]
```

Comment:
From how we want to test the stubs, this is not a good test, and I think it should be deleted.  What we want in the tests is the following:
1) Tests of valid code that should pass the type checker (that's about 98% of our tests)
2) Tests of invalid code that will raise exceptions that the type checker should catch.  We surround such tests with `if TYPE_CHECKING_INVALID_USAGE` and then put the `ignore` tags on those tests, which also means that the code will not be tested at runtime.

In your case, `bad_callable()` has a signature that could be a "good" callable, and the type checker can't tell the difference.  So  I don't think the 2 tests you have here should be included.  The test below is fine.

Reply:
I think the essence of these two tests is important. A callable that does not explicitly allow `pd.NA` as an argument should not be valid to be passed to `Series.map` if `na_action=None` (either explicitly or as a default value). The default value case was not detected before the changes that i made in this PR and caused a runtime error in a project of mine. So i do think it makes sense to explicitly test this case.

I do agree that they should be moved into a `TPYE_CHECHKING_INVALID_USAGE` block and have the pytest.raises removed.

Reply:
But then i would argue that the hints are just poor and that is purely on the user.

Take for example:
```
def add(a: float, b: float) -> float:
    return a+b

add(1.0, 2.0)
add("1", "2")
```
While both calls do not cause issues at run time, the second should definitely cause an error from the type checker. Because the type checker only validates the signature, because that is what external users interact with. The implementation could change at any point and cause runtime errors in the future if you are using the function in a way that does not match the signature.

even more explicitly
```
float_callable(pd.NA)
```
causes a typechecking failure by itself.

Reply:
> But then i would argue that the hints are just poor and that is purely on the user (author of the function).

I agree.  But we also make choices in the pandas stubs that cover the more common usage of pandas.  Handling the case of `np.nan` versus `pd.NA` is a long topic of discussion in the pandas team, and we hope to resolve it some day.  But right now, the majority of pandas users use `np.nan` to represent missing values, so we can't have stubs that prevent their code from working.  The changes you are proposing cause existing (and valid) pandas code to fail.

Reply:
But that point is only incidentally related to this, right? Because the current behavior is that only `Callable[[S1], S2 | NAType]` is allowed if the `na_type=None` (explicitely) and `Callable[[S1 | NAType], S2 | NAType]` for `na_type="ignore` (explicitely).

This PR mainly changes the case that not both cases are also accepted as the default value (which is only `None`).

If `np.nan` is commonly used as a nan value that the signature could be changed to `Callable[[S1], S2 | NAType | Literal[np.nan]]` and  `Callable[[S1 | NAType | Literal[np.nan]], S2 | NAType | Literal[np.nan]]` (probably also for the Mapping variants key and value types).

Because in the live version `s.map(bad_callable, na_type=None)` will give a type checking error.

Reply:
You could try adding the `Literal[np.nan]` in there, but I would want to make sure the following is accepted from a type checking perspective;
```python
s = pd.Series([1.1, 2.2, 3.3])
def good_callable(x: float) -> float:
    return x+ 1.3
s.map(good_callable, na_action=None)
s.map(good_callable)
s.map(good_callable, na_action="ignore")
```
If you can add that as a test, then I'd be OK with the change.

Reply:
I dont think the first should every typecheck (and it does not do so currently), because with the explicit `na_action=None` you are specifically saying that the function should be able to handle None-values, which `good_callable` does not do explicitly. And i feel that if you explicitly set this you probably do want to be warned about potential issues with your code (i library you are using could change the implementation of a function in a way that is non-breaking according to the signature and you could get runtime errors)

The third would work explicitly, while the second works currently but would not after this PR. I do concede that this case would probably cause warnings that currently work fine because the none values generally handle the common operations performed on numbers (+,-,*,/) without crashing (although importantly not all of them).

Another option here would be to try and carve out exceptions for int and float (maybe also complex and bool?) functions to not require the explicit none-value-handling because they support most of the operations performed on those, but leave it as is for the other (general case) because there they are likely to cause runtime errors?

Reply:
So if you want to separate the cases of `na_action=None` being explicit which should not be allowed, versus `na_action=None` being implicit (because `None` is the default value), I'd be OK with that.  For me, what's important is that the implicit case should pass type checking, because the default behavior (not specifying `na_action`) is most likely what people use today.

Reply:
Ok, that is basically exactly what the status is before this PR actually, because it allows the interpretation of `na_action="ignore"` as a default value which is more forgiving. Then i would just add a `TPYE_CHECKING_INVALID_USAGE` for `s.map(bad_callable, na_action=None)` which captures the current behavior.

---

Additionally, with `na_action=None` (especially implicitly), the reason why a lot of current incorrect (pure type theory wise) code doesnt cause runtime errors despite not handling `pd.NA` and `np.nan` explicitly, is because they play nicely with most number interactions by just propagating them. However, this is not the case with most other types. So maybe it would make sense to also fail on the implicit case for functions that take most types, but explicitly allow int/float/complex?

Ill try to see if i can do it and add it to the PR to show what i mean.

Reply:
Added the changes i proposed above and added some comments on the PR to explain them. Let me know what you think.

I actually also wanted to specifically allow mappings to return none-values `Mapping[S1, S2]` -> `Mapping[S1, S2 | NAType]` but this seems to mess with mypys overload resolution again.

Reply:
I'm OK with putting `# type: ignore` statements in the stubs when `mypy` and/or `pyright` complains about things like that.

Reply:
Yeah, but for mypy it also causes issues at the user side where it just infers things as `Any`.